### PR TITLE
FEATURE/calc-displacements

### PR DESCRIPTION
**Pull request by evamaxfield**
This is a branch off of base aics deformation object. So I will be closing that PR in favour of this one.

Highlights:
Multithreaded base calculate displacements
Multithreaded grid search for best parameters given seeds
Multithreaded full process displacements

Next up:
Streamline plots

### CODE REVIEW

File: aicsdeformation/aicsdeformation.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+class AICSDeformation(object):
+    """
+    AICSDeformation objects can be used to process and visualize standard deformation tasks.
+
+    :param frames: An ordered list of imaging frames, or a 3D ndarray, to compare.
+    :param t_index: If frames was provided as a 3D ndarray, which dimension index represents time.
+    """
+
+    @staticmethod
+    def _check_frames(frames: List[np.ndarray], dims: Tuple[int, int]):
```
Comment (heeler) [inline aicsdeformation/aicsdeformation.py:27]: I really like the typing here!

File: aicsdeformation/aicsdeformation.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+            if not isinstance(frame, np.ndarray):
+                raise TypeError("Frame data must be a numpy.ndarray.")
+
+            # Check frame is a 2D numpy.ndarray
+            if len(frame.shape) != 2:
+                raise ValueError("Frame data must be a 2D numpy.ndarray.")
+
+            # Check frame dimensions match
+            if frame.shape != dims:
+                raise ValueError("All frame data must have the same dimensions.")
```
Comment (heeler) [inline aicsdeformation/aicsdeformation.py:40]: Nice Sanity Checking!

File: aicsdeformation/aicsdeformation.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+                raise TypeError("No t_index was provided to split 3D ndarray into frames.")
+
+            # Split
+            frames = [frames[t, :, :] for t in range(frames.shape[t_index])]
+
+        # Check frames
+        self._check_frames(frames, frames[0].shape)
+
+        # Store frames
+        self._frames = [copy.deepcopy(frame.astype(np.uint16)) for frame in frames]
```
Comment (heeler) [inline aicsdeformation/aicsdeformation.py:60]: If this were C++ I would make _frames static meaning that once it was assigned it was immutable and that any function that took it as an argument would have to preserve the static nature of the data. I don't think python has an equivalent so making a copy is the best within the language but if it becomes performance prohibitive maybe we need to think about this more.

File: aicsdeformation/aicsdeformation.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        # Lazy load
+        self._displacements = None
+
+    @property
+    def frames(self):
+        return self._frames
+
+    @property
+    def dims(self):
+        return self[0].shape
```
Comment (heeler) [inline aicsdeformation/aicsdeformation.py:71]: Does dims just return the X, Y shape or am I missing something here?

File: aicsdeformation/aicsdeformation.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+    def insert_frames(self, index: int, frames: List[np.ndarray]):
+        """
+        Insert multiple frames.
+
+        :param index: Which index to insert the frames at.
+        :param frames: List of frames to insert.
+        """
+        # Check frames
+        self._check_frames(frames, self.dims)
```
Comment (heeler) [inline aicsdeformation/aicsdeformation.py:81]: I really like how you've got the _check_frames function. The design of this is very nice.

File: aicsdeformation/aicsdeformation.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+
+        :param index: Which frame index to update.
+        :param frame: New frame data to set at the index, if None, the frame at target index will be removed.
+        :return: The previously stored frame data at the provided index.
+        """
+        # Check frame
+        self._check_frames([frame], self.dims)
+
+        # Insert and remove old
+        self.insert_frame(index, frame)
```
Comment (heeler) [inline aicsdeformation/aicsdeformation.py:146]: What happens if the insert fails?

File: aicsdeformation/aicsdeformation.py [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+        return self.remove_frame(index + 1)
+
+    @property
+    def displacements(self):
+        if self._displacements is None:
+            self._displacements = calculate_displacements(self)
+
+        return self._displacements
+
+    def generate_displacements(
```
Comment (heeler) [inline aicsdeformation/aicsdeformation.py:156]: Are we better off using a named tuple to contain this?

File: aicsdeformation/aicsdeformation.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+
+        Additionally, a grid search implementation is available in the case you want to have aicsdeformation
+        find the the best signal : noise given starting parameters, example below.
+
+        ```
+        from aicsdeformation.processing import grid_search_displacements
+        best, all_displacements = grid_search_displacements(frame_a, frame_b)
+        ```
+        """
+        # Get base
```
Comment (heeler) [inline aicsdeformation/aicsdeformation.py:185]: Thanks! You've cleaned this up really nicely!

File: aicsdeformation/processing.py [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+        This will be expanded to frame_a and frame_b if not None.
+    :param frame_a: A single frame to be used to find displacements for.
+    :param frame_b: A single frame to be used to find displacements for.
+
+    Look to OpenPIV for details on how parameters interact with the actual displacement generation.
+    https://openpiv.readthedocs.io/en/latest/src/tutorial.html
+    """
+    # Expand frames if required
+    if frames:
+        frame_a, frame_b = (*frames,)
```
Comment (heeler) [inline aicsdeformation/processing.py]: Is the comma in (*frames,) just to protect against frames only containing a single frame?

File: aicsdeformation/processing.py [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+    return best_dfp, dfps
+
+
+def process_displacement(
+    d: Displacement,
+    image_size: Tuple[int],
+    window_size: int = 16,
+    overlap: int = 4,
+    s2n_threshold: float = 1.3,
+    outlier_method: str = "localmean",
```
Comment (heeler) [inline aicsdeformation/processing.py:267]: Nice I like that the outlier method is a parameter that can be changed.

File: aicsdeformation/types.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+import numpy as np
+
+
+class Displacement(object):
```
Comment (heeler) [inline aicsdeformation/types.py:7]: This is great. I usually do this in C++ and should have thought to do it here.

**Review by heeler — COMMENTED** [review]
I really like what you've done. The few spots where I've made comments are likely due to a lack of knowledge on my part.

File: aicsdeformation/aicsdeformation.py (hunk 3 above)
Comment (evamaxfield) [reply aicsdeformation/aicsdeformation.py:60]: The equivalent is the `_`. While still accessible to users, generally the agreement is to not touch underscore methods or attributes. If we wanted to make it truly invisible I believe you can `__`, not sure though.

File: aicsdeformation/aicsdeformation.py (hunk 4 above)
Comment (evamaxfield) [reply aicsdeformation/aicsdeformation.py:71]: Nope it really just returns the X, Y shape. I just didn't like typing `self[0].shape` every time I wanted the dims.

File: aicsdeformation/aicsdeformation.py (hunk 6 above)
Comment (evamaxfield) [reply aicsdeformation/aicsdeformation.py:146]: If the insert fails, something has gone drastically wrong with basic python array insert. As really all of the insert, update, remove functions and their variants are just wrappers around python list manipulation.

File: aicsdeformation/aicsdeformation.py (hunk 7 above)
Comment (evamaxfield) [reply aicsdeformation/aicsdeformation.py:156]: You mean a named tuple for parameters? Not sure I follow

File: aicsdeformation/processing.py (hunk 9 above)
Comment (evamaxfield) [reply aicsdeformation/processing.py]: This made me realize I don't need to do this. I can instead just do `a, b = frames` and it is equivalent....

File: aicsdeformation/aicsdeformation.py (hunk 3 above)
Comment (heeler) [reply aicsdeformation/aicsdeformation.py:60]: I said static I meant const. I get the _ idea, in C/C++ the compiler enforces that contract meaning the code won't compile unless the functions are declared const or take const arguments.

**Pull request merged**